### PR TITLE
Use minimum of memory limits for external aggregation/sort

### PR DESCRIPTION
**Pull request by Michicosun**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory
and will run independently of the checks below:

- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64
- [ ]  Exclude: All with release
- [ ]  Exclude: All with debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 1f23acc1f7f8b3086fea446741b2d90147ba2c84 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73933/1f23acc1f7f8b3086fea446741b2d90147ba2c84/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/73836/a0cb555aa4aa6eef3510d7502aa427688033677f/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/Aggregator.cpp [hunk 1]
```diff
@@ -250,7 +255,10 @@ size_t Aggregator::Params::getMaxBytesBeforeExternalGroupBy(size_t max_bytes_bef
         }
     }
 
-    return max_bytes_before_external_group_by;
+    if (threshold == std::numeric_limits<size_t>::max())
```
Comment (alexey-milovidov) [inline src/Interpreters/Aggregator.cpp:258]: This roundabout looks tricky. You can use std::optional instead.

Comment (Michicosun) [reply src/Interpreters/Aggregator.cpp]: Good point, changed to optional

**Review by alexey-milovidov — APPROVED** [review]

**Comment by rschu1ze** [comment]
@Michicosun This PR deserves a changelog entry.

**Comment by alexey-milovidov** [comment]
@rschu1ze, this is how it is supposed to work in the first place in 24.12. I'll add "must backport".

**Pull request merged**